docs/content/_index.md

```diff
@@ -1,608 +1,1007 @@
+You can store values in intermediate variables using the `:=` (assignment)
+operator. Variables can be referenced just like `input`.
 
-### Rules and Virtual Documents
+```live:example/vars/1:query:merge_down
+s := input.servers[0]; s.id == "app"; p := s.protocols[0]; p == "https"
+```
+```live:example/vars/1:output
+```
 
-In contrast to base documents, virtual documents embody the results of evaluating the rules included in policy modules. Virtual documents are computed when users publish new policy modules, update existing modules, run queries, and when any relevant base document is published or updated. Rules allow policy authors to write questions with yes-no answers (that is, predicates) and to generate structured values from raw data found in base documents as well as from intermediate data found in other virtual documents.
+When OPA evaluates expressions, it finds values for the variables that make all
+of the expressions true. If there are no variable assignments that make all of
+the expressions true, the result is undefined.
 
-### The `data` Document
+```live:example/vars/undefined:query:merge_down
+s := input.servers[0]; s.id == "app"; s.protocols[1] == "telnet"
+```
+```live:example/vars/undefined:output:expect_undefined
+```
 
-All documents pushed into OPA or computed by rules are nested under a built-in root document named data.
+Variables are immutable. OPA reports an error if you try to assign the same
+variable twice.
 
-{{< figure src="data-model-logical.svg" width="70" caption="OPA document structure" >}}
+```live:example/vars/assign_error:query:merge_down
+s := input.servers[0]; s := input.servers[1]
+```
+```live:example/vars/assign_error:output:expect_assigned_above
+```
 
-Example `data` document:
+OPA must be able to enumerate the values for all variables in all expressions.
+If OPA cannot enumerate the values of a variable in any expression, OPA will
+report an error.
 
-```json
-{
-  "servers": [...],
-  "ports": [...],
-  "networks": [...],
-  "opa": {
-    "examples": {
-      "violations": [...],
-      "public_servers": [...]
-    }
-  }
+```live:example/vars/unsafe:query:merge_down
+x := 1; x != y  # y has not been assigned a value
+```
+```live:example/vars/unsafe:output:expect_unsafe_var
+```
+
+### Iteration
+
+```live:example/iter:module:hidden
+package example
+```
+
+Like other declarative languages (e.g., SQL), Rego does not have an explicit
+loop or iteration construct. Instead, iteration happens implicitly when you
+inject variables into expressions.
+
+To understand how iteration works in Rego, imagine you need to check if any
+networks are public. One option would be to test each network in the input:
```

Comment:
One thing I noticed while reading through the tutorial was that by about this point in time I didn't remember what the input was exactly and had to scroll back up to the top to look at the structure so I could play around with the example (subsequently realizing I could just evaluate `input.networks` or something to see).

I'm not sure I have a great idea to help with this aside from maybe making another "pro tip" note somewhere on here to explain how to explore the input object when looking at the examples more. Or maybe before something like jumping into the `input.networks[0].public == true` example show the base object first like `input.networks[0]` and then show the boolean query.

Reply:
Done. I've added a query to show the networks before diving into the section.